File: python/sorting/merge_sort.py
def merge_sort(arr):
    """
    Sorts an array using the merge sort algorithm.

    :param arr: List of elements to be sorted.
    :return: New list of sorted elements.
    """
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2

    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left,right)

def merge(left, right):
    """
    Merges two sorted lists into a single sorted list.

    :param left: First sorted list.
    :param right: Second sorted list.
    :return: Merged and sorted list.
    """
    sorted_arr = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_arr.append(left[i])
            i = i+1
        else:
            sorted_arr.append(right[j])
            j = j+1
    
    sorted_arr.extend(left[i:])
    sorted_arr.extend(right[j:])

    return sorted_arr

if __name__ == "__main__":
    print("Merge Sort Simple Test")
    arr = [7,3,2,4,6]
    print(arr)
    output = merge_sort(arr)
    print(output)

File: csharp/sorting/MergeSort.cs
using System.ComponentModel.DataAnnotations;
using Sorting;

namespace Algorithms.Sorting
{
    public class MergeSort<T> : IComparisonSorting<T> where T : IComparable<T>
    {
        public void Sort(T[] arr)
        {
            MergeSortAlgo(arr, 0, arr.Length-1);
        }

        private void MergeSortAlgo(T[] arr, int left, int right) 
        {
            if(left == right)
                return;
            
            int m = (int)((left + right) / 2);
            MergeSortAlgo(arr, left, m);
            MergeSortAlgo(arr, m+1, right);
            Merge(arr, left, right);
        }

        private void Merge(T[] arr, int left, int right) 
        {
            T[] tmp = new T[right-left+1];
            int m = (left + right) / 2;
            int w = 0, i = left, j = m+1;
            while (i <= m && j <= right) 
            {
                if(arr[i].CompareTo(arr[j]) < 0) 
                    tmp[w++] = arr[i++];
                else
                    tmp[w++] = arr[j++];
            }
            
            while (i <= m) tmp[w++] = arr[i++];
            while (j <= right) tmp[w++] = arr[j++];
            tmp.CopyTo(arr, left);
        }

    }
}
